src/error.rs

```diff
@@ -1,85 +1,151 @@
 use std::fmt;
 
-use reqwest::Method;
+use reqwest::{Method, Response};
+use serde::{Deserialize, Serialize};
 
 #[derive(Debug)]
-pub struct HttpError {
-    inner: InnerHttpError,
-    method: Method,
-    path: String,
+pub enum ApiError {
+    GatewayDeleted,
 }
 
-impl HttpError {
-    pub fn request_method(&self) -> &Method {
-        &self.method
-    }
+// Response from server
+#[derive(Debug, Deserialize, Serialize)]
+struct ApiServerResponse {
+    code: String,
+}
 
-    pub fn request_path(&self) -> &str {
-        &self.path
+impl<'de> Deserialize<'de> for ApiError {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use heck::KebabCase;
+        let resp = ApiServerResponse::deserialize(deserializer)?;
+        match resp.code.to_kebab_case().as_str() {
+            "gateway-deleted" => Ok(ApiError::GatewayDeleted),
+            _ => Err(serde::de::Error::custom(format!("Couldn't deserialize a: {:?}", resp))),
+        }
     }
+}
 
-    /// If the error comes from `reqwest`, return the inner `reqwest::Error`.
-    pub fn to_reqwest_error(&self) -> Option<&reqwest::Error> {
-        match &self.inner {
-            InnerHttpError::Url(_) | InnerHttpError::Query(_) => None,
-            InnerHttpError::Reqwest(e) => Some(e),
+impl std::fmt::Display for ApiError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ApiError::GatewayDeleted => write!(f, "ApiError : GatewayDeleted"),
```

Comment:
`Display` is for human-readable messages, can we use the full error message here?

```suggestion
            ApiError::GatewayDeleted => write!(f, "The gateway this access token belonged to has been deleted"),
```

Reply:
Do we need `Display` at all if we have a human-readable `message` from the server?

Reply:
Yes, we need it for the type to be able to implement `std::error::Error`. It will likely be used for logs.